File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int left = 0;
        int right = n - 1;
        int water = 0;
        int lmax = 0;
        int rmax = 0;

        while (left < right) {

            lmax=Math.max(height[left],lmax);
            rmax=Math.max(height[right],rmax);

            if (lmax < rmax) {
                water+= lmax -height[left];
                left++;
            }
            else{
                water+=rmax-height[right];
                right--;
            }
        }
        return water;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        lmax = 0
        rmax = 0
        left = 0
        right = len(height) - 1
        water = 0

        while left <= right:
            lmax = max(height[left], lmax)
            rmax = max(height[right], rmax)

            if lmax <= rmax:
                water += lmax - height[left]
                left += 1
            elif rmax < lmax:
                water += rmax - height[right]
                right -= 1

        return water
